### PR TITLE
[CN-369] WAN Sync

### PR DESCRIPTION
**Pull request by SeriyBg**
PR closed by Hazelcast automation as no activity (>3 months). Please reopen with comments, if necessary. Thank you for using Hazelcast and your valuable contributions

### CODE REVIEW

File: controllers/hazelcast/t/main.go [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package main
```
Comment (SeriyBg) [inline controllers/hazelcast/t/main.go:1]: This one is accidentally here, I will remove it

File: controllers/hazelcast/wansync_controller.go [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+			return updateWanSyncStatus(ctx, r.Client, wan, wanSyncFailedStatus(err))
+		} else {
+			return updateWanSyncStatus(ctx, r.Client, wan, wanSyncPendingStatus())
+		}
+	}
+
+	// Check publisherId is registered to the status, otherwise issue WanReplication config to Hazelcast
+	if wan.Status.PublisherId == "" {
+		logger.Info("Applying WAN configuration")
+		publisherId, err := r.getWanPublisherId(ctx, cli, wan)
```
Comment (mtyazici) [inline controllers/hazelcast/wansync_controller.go:82]: My question is does this not start a Wan Replication right away if they give the config in WanSync? And then we send a wan sync request later, what happens in between. Up until this point it seems like what we are doing is the same as Wan Replication.

**Review by mtyazici — COMMENTED** [review]
Thanks for the PR, I have some questions regarding to the WanSync and reconciler logic order. Also left a few tirivial comments.

File: controllers/hazelcast/hazelcast_rest_client.go [hunk 3]
```diff
@@ -137,6 +139,47 @@ func (c *RestClient) HotBackup(ctx context.Context) error {
...
+	if s := rBody["status"]; s != "success" {
+		return fmt.Errorf("error executing WAN Sync: %s, %s", s, rBody["message"])
+	}
+	return nil
+}
+
+func (c *RestClient) wanSyncData(wan WanPublisherObject) string {
+	wpc := wan.WanPublisherConfig()
+	if wpc.MapResourceName != "" {
+		return fmt.Sprintf("%s&&%s&%s&%s",
```
[the remark this answers is not in the document]
Comment (SeriyBg) [reply controllers/hazelcast/hazelcast_rest_client.go:175]: Yeap, this is for password which is not supported for now in operator

File: controllers/hazelcast/wansync_controller.go (hunk 2 above)
Comment (SeriyBg) [reply controllers/hazelcast/wansync_controller.go:82]: That's true. For WanSync we need to use a WAN publisher. The only difference is that the Sync operation will send the existing events to the target cluster, while the replication is only working with the new events

**Pull request closed without merging**